### PR TITLE
lesson 3 tasks

### PR DESCRIPTION
**Pull request by AZhurav1ev**
Completed lesson 3 tasks, reviewed some functions from privious lessons.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -14,13 +14,25 @@
...
         
         .green {
             color: green;
         }
+        
+        .gray {
+            background: #F4F5F9;
+        }
+        
+        .myhover {
```
Comment (roman-zaitsev) [inline index.html]: bad class name

File: package-lock.json [hunk 2]
```diff
@@ -0,0 +1,192 @@
+{
```
Comment (roman-zaitsev) [inline package-lock.json]: package-lock.json without package.json

File: index.html (hunk 1 above)
Comment (AZhurav1ev) [reply index.html]: Renamed

File: package-lock.json (hunk 2 above)
Comment (AZhurav1ev) [reply package-lock.json]: Added package.json

File: script.js [hunk 3]
```diff
@@ -1,25 +1,28 @@
...
-    const validationResult = $$("form").validate();
-    if (validationResult) {
-        const item_data = $$("form").getValues();
-        $$("table").add(item_data);
-        $$("form").clear();
+    const form = $$("form");
+    const table = $$("table");
+    const item_data = form.getValues();
+    if (item_data.id) {
+        table.updateItem(item_data.id, item_data);
```
Comment (roman-zaitsev) [inline script.js]: missing validation before update

File: script.js [hunk 4]
```diff
@@ -1,25 +1,28 @@
...
+        form.clear();
```
Comment (roman-zaitsev) [inline script.js]: after clear form table select is still exists, use unselect for table

File: script.js [hunk 5]
```diff
@@ -75,9 +89,27 @@ const list = {
...
+    on: {
+        onAfterSelect: dataToForm
+    },
+    columns: [
+        { id: "rank", header: "", width: 50, css: "gray" },
+        { id: "title", header: ["Film Title", { content: "textFilter" }], fillspace: true },
+        { id: "year", header: ["Released", { content: "numberFilter" }] },
+        { id: "votes", header: ["Votes", { content: "textFilter" }] },
+        { id: "rating", header: ["Rating", { content: "textFilter" }] },
+        { id: "deleteIcon", header: "", template: "{common.trashIcon()}" }
```
Comment (roman-zaitsev) [inline script.js]: Add sorting and filtering to all Datatable columns (as in the picture below)

there is no sorting for columns

File: script.js [hunk 6]
```diff
@@ -128,18 +160,110 @@ const footer = {
...
+            elements: [
+                {
+                    view: "text",
+                    id: "title_filter_input",
+                    placeholder: "Country name",
+                    on: {
+                        onTimedKeyPress: function () {
+                            const value = this.getValue().toLowerCase();
+                            $$("user_list").filter(function (obj) {
+                                return obj.country.toLowerCase().indexOf(value) !== -1;
```
Comment (roman-zaitsev) [inline script.js]: I'm confused @Zwillinge 
Should it filter only by country? The task is "Enable List filtering via a Text input"

File: script.js [hunk 7]
```diff
@@ -128,18 +160,110 @@ const footer = {
...
+const treeTable = {
+    view: "treetable",
+    id: "price_tree",
+    columns: [
+        { id: "id", header: "", width: 50 },
+        { id: "title", header: "Title", template: "{common.treetable()} #title#", width: 200 },
+        { id: "price", header: "Price", width: 200 }
+    ],
+    autowidth: true,
+    select: true,
```
Comment (roman-zaitsev) [inline script.js]: The task is "Enable cell selection"

File: script.js (hunk 6 above)
Comment (Zwillinge) [reply script.js]: it should be filtered by name or by both name and country (visible values)

File: script.js (hunk 3 above)
Comment (Zwillinge) [reply script.js]: it should be done like this: 

- validation
- getting values and reading `id`
- add or update based on the existence of the id field

Comment (AZhurav1ev) [reply script.js]: There was an issue in tranformering ratings and votes values before verification. But now it's fixed.

File: script.js (hunk 4 above)
Comment (AZhurav1ev) [reply script.js]: Added this metod.

File: script.js (hunk 5 above)
Comment (AZhurav1ev) [reply script.js]: Didn't notice it. Fixed.

File: script.js (hunk 6 above)
Comment (AZhurav1ev) [reply script.js]: Added additional filter for names.

File: script.js (hunk 7 above)
Comment (AZhurav1ev) [reply script.js]: Fixed.

File: script.js [hunk 8]
```diff
@@ -1,25 +1,29 @@
...
-            expire: 1000
-        });
+    const form = $$("form");
+    const table = $$("table");
+
+    if (form.validate()) {
+        const item_data = form.getValues();
+        if (item_data.id) {
+            table.updateItem(item_data.id, item_data);
+            form.clear();
```
Comment (Zwillinge) [inline script.js]: form.clear is a common action, it's better to place it after if..else block

File: script.js [hunk 9]
```diff
@@ -75,9 +90,37 @@ const list = {
...
+            this.remove(id.row)
+        }
+    },
+    scheme: {
+        $init: function (obj) {
+            if (obj.votes.includes(",")) {
+                obj.votes = obj.votes.replace(",", ".") * 1000;
+            }
+            if (obj.rating.includes(",")) {
+                obj.rating = obj.rating.replace(",", ".");
```
Comment (Zwillinge) [inline script.js]: by this `replace` you actually decrease ratings by 1000. why not simply replace commas by empty strings?

File: script.js (hunk 8 above)
Comment (AZhurav1ev) [reply script.js]: Thank you, moved it out from if...else block.

File: script.js (hunk 9 above)
Comment (AZhurav1ev) [reply script.js]: "All genius is simple", fixed it.

**Pull request merged**